binfmt/binfmt_exec.c

```diff
@@ -121,7 +121,7 @@ int exec_spawn(FAR const char *filename, FAR char * const *argv,
 
   /* Then start the module */
 
-  pid = exec_module(bin, filename, argv);
+  pid = exec_module(bin, filename, argv, NULL);
```

Comment:
always fail at line 157

Reply:
I don't understand, what is the failure ?

Reply:
Ok, you don't really copy envp directly in binfmt_copyenv, but from group->tg_envp, which confuse me. If so why do you pass envp to binfmt_copyenv?

Reply:
envp should come all the way from posix_spawn, but it does not. This just takes passing the argument one step closer to where it should be. Not a required change but does not create complexity or break anything, I think this is fine.

Reply:
How about we add envp to all execution function? It isn't too hard with #6083 now.

Reply:
My worry is that we break some existing applications with very low reason to do so. But to me it is fine to pass it all the way.

Reply:
Done, at least partially. The last mile, how to provide char **environ; to /apps is still open. I can spend a short while trying to figure out how to do it, or we can leave this partial solution as is and fix that later.

posix_spawn() already passes envp[] which is simply set to NULL always, the same thing can be done for execve() and so forth.

Reply:
> Done, at least partially. The last mile, how to provide char **environ; to /apps is still open. I can spend a short while trying to figure out how to do it, or we can leave this partial solution as is and fix that later.

it's here:
https://github.com/apache/incubator-nuttx/blob/master/include/stdlib.h#L61-L65
one goal of #6083 is to support environ fully:
https://github.com/apache/incubator-nuttx/pull/6083/files#diff-a5924166288d58f423d0ff30ddd13c61f637ca82b180ce3f0b2aefda056ba323R55-R60

> 
> posix_spawn() already passes envp[] which is simply set to NULL always, the same thing can be done for execve() and so forth.

Reply:
Nice, I did not notice it was already done. Should be quite trivial to fix it then.

Reply:
I will make the modification to nsh_fileapps and try to figure out how to add call gate for get_environ_ptr() (it obviously cannot be called directly in kernel mode).

Reply:
Done